specification/transport_layer/can.tex

```diff
@@ -208,16 +207,15 @@ \subsubsection{Source node-ID field in anonymous message transfers}\label{sec:ca
 CAN bus does not allow different nodes to transmit CAN frames with different data field values under the same CAN ID.
 Owing to the fact that the CAN ID includes the node-ID value of the transmitting node,
 this restriction does not affect regular UAVCAN transfers.
-However, anonymous message transfers would violate this restriction,
-because they don't have a unique node-ID.
+However, anonymous message transfers would violate this restriction because they don't have a unique node-ID.
 
 In order to work around this problem,
 UAVCAN requires that the source node-ID field of anonymous messages\footnote{Source node identifier
 is not defined for anonymous message transfers; see table \ref{table:common_transfer_properties}.}
 is initialized with a pseudorandom \emph{pseudo-ID value},
 and defines special logic for handling CAN bus errors during transmission of anonymous frames.
 
-The source of the pseudorandom data used for the pseudo-ID must be likely to produce different values
+The source of the pseudorandom data used for the pseudo-ID must be likely to produce different values % Not sure what 'must be likely' means in this context'
```

Comment:
Poor wording. What I am trying to say here is that the following conditions apply:

- Identical pseudo-ID values given different payloads are undesirable.
- Identical pseudo-ID values given identical payload are acceptable.

Perhaps the sentence should be changed to communicate these objectives?

Reply:
Identical pseudo-ID given different payloads given simultaneous transmission triggers a collision error on the bus. Collision is bad because it forces all nodes involved in the collision to re-transmit, delaying their progress and temporarily disrupting all lower-priority traffic on the bus.

The anonymous transfer logic is fundamentally non-deterministic, but it doesn't mean that we don't care about its performance. Hence, we want to minimize the probability of collision, hence the objective to make the pseudo-ID sufficiently random.

Reply:
Thanks for the explanation! I'm guessing that's already explained elsewhere in the document or known by whoever will implement this? 

If that's the case, I'm thinking this can just be rephrased to "The source of the pseudorandom data used for the pseudo-ID should ideally produce different values for different payloads. "

Reply:
> I'm guessing that's already explained elsewhere in the document or known by whoever will implement this?

It is implied that the reader is already familiar with the CAN bus specification (it is an ISO standard), where this is explained.

> The source of the pseudorandom data used for the pseudo-ID should ideally produce different values for different payloads.

I am certain that everyone will understand what we mean, but I am not exactly sure about the technical correctness here. The number of distinct pseudo-ID values is less than the number of distinct payload values; therefore (pigeonhole principle) the ideal source of pseudo-ID, as we have defined it, cannot exist. The words "must be likely" in my original text were intended to address this problem, but I see now that they hardly added any clarity here. 

Maybe we could say something along these lines instead:

> The source of the pseudorandom data used for the pseudo-ID should be chosen so that the following principles are accommodated:
>
> - Identical pseudo-ID values given different payloads are undesirable.
> - Identical pseudo-ID values given identical payload are acceptable.

Or maybe I am just paying too much attention or otherwise overreacting. What do you think? What does @thirtytwobits think?

Reply:
I think what's confusing, this being part of a specification, is "must" coupled with "likely". The two concepts conflict (You MUST press the green button to avoid self-destruction since it is LIKELY that pressing the green button will cancel the self-destruct activity). We should change one of the two terms:

> If source of the pseudorandom data used for the pseudo-ID is likely to produce duplicate values more anonymous messages will be lost.

or 

> The source of the pseudorandom data used for the pseudo-ID must produce different values or more anonymous message will be lost.

Reply:
> The source of the pseudorandom data used for the pseudo-ID must produce different values **for different CAN frame payloads** or more anonymous messages will be lost.

The part about payloads seemed missing.

*(I don't see any contradiction in the green button example. Assuming that your self-destruction is certain unless the button is pressed; the chances of self-destruction, if the button is pressed, are less than certain; and pressing the button does not cause any unexpected side-effects, you MUST press the button (assuming that self-destruction is not your preferred outcome)).*